### ISSUE TITLE
DoRA slow forward inference

### ISSUE THREAD

**Issue by phemw** [huggingface/peft#2651]
### System Info

Please update the forward of DoRA so that during evaluation it doesn't wastefully recompute weight normalization during each forward pass when it doesn't need to. Hopefully switching between the current implementation and an implementation that precomputes the weight norm is as simple as checking `model.eval() # not model.training`.

I understand one could merge the weights but it tends to be more lossy than simple unmerged adapters. Additionally, this should speed up evals during finetuning larger models.

### Who can help?

@BenjaminBossan 

### Reproduction

https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/dora.py#L66-L101

### Expected behavior

Inference with unmerged adapters (`model.training==False`) does not compute weight norm during each forward pass.

**Comment by phemw** [comment]
Hello any updates on this @BenjaminBossan? Also what is the difference between [peft/tuners/lora/dora.py](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/dora.py) and [peft/tuners/lora/variants.py](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/variants.py), thanks.

**Comment by BenjaminBossan** [comment]
@phemw I was at a conference, hence the late reply.

You are right that during inference, the weight norm of DoRA is needlessly recomputed each time. I created a draft PR to cache the value: #2661. Please check my description there. I assume that you have a real world use case, it would be great if you could measure the speed benefits of that branch against this use case and report the outcome.

> Also what is the difference between [peft/tuners/lora/dora.py](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/dora.py) and [peft/tuners/lora/variants.py](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/variants.py), thanks.

This is more of an implementation detail. We introduced LoRA variants as a low friction way to add new LoRA-like PEFT methods like DoRA. These variant classes provide the interface with the LoRA layer. The modules defined in `dora.py` itself contain the actual DoRA implementation for the layers.

**Comment by phemw** [comment]
Hi saw the PR, I think you should get some noticeable difference on models larger than `meta-llama/Llama-3.2-1B`. For instance, on `meta-llama/Llama-3.1-8B` LoRA takes ~20s vs. DoRA ~100s to generate 1k tokens, and I presume caching the norm should close this gap. If you don't get similar results I am happy to diagnose further, thanks.

**Comment by BenjaminBossan** [comment]
I tried `meta-llama/Llama-3.1-8B` and the difference is bigger, but still small overall (16%). It would be great if you could test on your use case, so that we can identify where the difference stems from.

For reference, here is the script I used to test it:

```python
from contextlib import contextmanager
import time

import torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model

device = 0
num_runs = 10
model_id = "meta-llama/Llama-3.1-8B"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, device_map=device)
config = LoraConfig(init_lora_weights=False, use_dora=True)
model = get_peft_model(model, config)

durations = {
    "eval": [],
    "train": [],
}

@contextmanager
def timeit(name):
    start = time.perf_counter()
    yield
    end = time.perf_counter()
    dur = end - start
    print(f"{name:>5} mode took {dur:.4f} seconds")
    durations[name].append(dur)

for _ in range(num_runs + 1):
    model.eval();
    with timeit("eval") as dur:
        for i in range(3):
            x = torch.randint(10, 100, (1, 50)).to(device)
            model(x)

    model.train();
    with timeit("train") as dur:
        for i in range(3):
            x = torch.randint(10, 100, (1, 50)).to(device)
            model(x)

# remove the first run, which is usually slower due to model initialization
del durations["eval"][0]
del durations["train"][0]

print("summary")
for name, dur in durations.items():
    print(f"{name}: {num_runs} runs, avg {sum(dur)/num_runs:.4f} sec, min {min(dur):.4f} sec, max {max(dur):.4f} sec")
```

which gives me:

```
eval: 10 runs, avg 0.1353 sec, min 0.1348 sec, max 0.1355 sec
train: 10 runs, avg 0.1612 sec, min 0.1609 sec, max 0.1616 sec
```

(note: times are much faster than previous reported result because it's on GPU this time)

PS: Can you share which base model you're using and which layers you're targeting?

**Comment by phemw** [comment]
Hi the difference between train and eval are small, which is good and expected in this case, since the same amount of forward passes are used.

However the difference between these two are quite drastic (not running https://github.com/huggingface/peft/pull/2661 code) .

```python
config1 = LoraConfig(init_lora_weights=False, use_dora=True)
config2 = LoraConfig(init_lora_weights=False, use_dora=False)
```

Roughly 170% based on your script. Let me know if you get similar results.

**Comment by BenjaminBossan** [comment]
Yes, turning DoRA off completely gives a difference of that magnitude. But that just shows that DoRA adds a general overhead, not that the weight norm computation itself is at fault, right?

I turned on `line_profiler` for this script and profiled the DoRA `forward` step, giving me:

```python
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    68                                               def forward(self, x, *, lora_A, lora_B, scaling, base_layer, base_result=None):                                    
    69                                                   """                                                                                                            
    70                                                   For DoRA, calculate the extra output from LoRA with DoRA applied. This should be added on top of the base layer
    71                                                   output.                                                                                                        
    72                                                   """                                                                                                            
    73                                                   # Don't use `lora_weight = lora_B.weight @ lora_A.weight` because this causes errors with FSDP. Instead,       
    74                                                   # calculate the same but using forward.                                                                        
    75      4224     167671.1     39.7     15.8          x_eye = torch.eye(lora_A.weight.shape[1], device=lora_A.weight.device, dtype=x.dtype)                          
    76      4224     330127.7     78.2     31.2          lora_weight = lora_B(lora_A(x_eye)).T                                                                          
    77                                                                                                                                                                  
    78      4224       3664.6      0.9      0.3          magnitude = self.weight                                                                                        
    79      4224      21692.4      5.1      2.0          weight = dequantize_module_weight(base_layer)                                                                  
    80      4224      50835.1     12.0      4.8          weight = weight.to(x.dtype)                                                                                    
    81      4224     170762.2     40.4     16.1          weight_norm = self.get_weight_norm(weight, lora_weight.detach(), scaling)                                      
    82                                                   # see section 4.3 of DoRA (https://huggingface.co/papers/2402.09353)                                           
    83                                                   # "[...] we suggest treating ||V +∆V ||_c in                                                                   
    84                                                   # Eq. (5) as a constant, thereby detaching it from the gradient                                                
    85                                                   # graph. This means that while ||V + ∆V ||_c dynamically                                                       
    86                                                   # reflects the updates of ∆V , it won’t receive any gradient                                                   
    87                                                   # during backpropagation"                                                                                      
    88      4224       4025.5      1.0      0.4          weight_norm = weight_norm.detach()                                                                             
    89      4224      40906.1      9.7      3.9          mag_norm_scale = (magnitude / weight_norm).view(1, -1)                                                         
    90                                                                                                                                                                  
    91      4224     141684.7     33.5     13.4          lora_result = lora_B(lora_A(x))                                                                                
    92                                                                                                                                                                  
    93      4224        497.6      0.1      0.0          bias = None                                                                                                    
    94      4224        599.7      0.1      0.1          if base_result is not None:                                                                                    
    95      4224       2758.6      0.7      0.3              bias = base_layer.bias                                                                                     
    96      4224        467.8      0.1      0.0              if bias is not None:                                                                                       
    97                                                           base_result = base_result - bias                                                                       
    98                                                   else:                                                                                                          
    99                                                       base_result = F.linear(x, transpose(weight, self.fan_in_fan_out))                                          
   100                                                                                                                                                                  
   101      4224     120809.3     28.6     11.4          result_dora = (mag_norm_scale - 1) * base_result + mag_norm_scale * lora_result * scaling                      
   102                                                                                                                                                                  
   103      4224       1895.4      0.4      0.2          return result_dora                                                                                             
```

So from the looks of it, the weight norm computation takes only about 16%, which is consistent with the 16% speed increase reported above when enabling caching. There is no single line here that could be easily optimized for a significant performance gain.

**Comment by phemw** [comment]
Thanks, from the stack trace:

```python
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================                                                                   
    75      4224     167671.1     39.7     15.8          x_eye = torch.eye(lora_A.weight.shape[1], device=lora_A.weight.device, dtype=x.dtype)                          
    76      4224     330127.7     78.2     31.2          lora_weight = lora_B(lora_A(x_eye)).T                                                                          
    77                                                                                                                                                                  
    78      4224       3664.6      0.9      0.3          magnitude = self.weight                                                                                        
    79      4224      21692.4      5.1      2.0          weight = dequantize_module_weight(base_layer)                                                                  
    80      4224      50835.1     12.0      4.8          weight = weight.to(x.dtype)                                                                                    
    81      4224     170762.2     40.4     16.1          weight_norm = self.get_weight_norm(weight, lora_weight.detach(), scaling)                                      
    82                                                   # see section 4.3 of DoRA (https://huggingface.co/papers/2402.09353)                                           
    83                                                   # "[...] we suggest treating ||V +∆V ||_c in                                                                   
    84                                                   # Eq. (5) as a constant, thereby detaching it from the gradient                                                
    85                                                   # graph. This means that while ||V + ∆V ||_c dynamically                                                       
    86                                                   # reflects the updates of ∆V , it won’t receive any gradient                                                   
    87                                                   # during backpropagation"                                                                                      
    88      4224       4025.5      1.0      0.4          weight_norm = weight_norm.detach()                                                                             
    89      4224      40906.1      9.7      3.9          mag_norm_scale = (magnitude / weight_norm).view(1, -1)                                                         
```

I suppose all of this (except lines [79, 80]) could be precomputed and that should close the gap during unmerged forward passes.

**Comment by BenjaminBossan** [comment]
> I suppose all of this (except lines [79, 80]) could be precomputed and that should close the gap during unmerged forward passes.

Not really, everything that involves `x` itself cannot be precomputed, i.e. everything from here:

https://github.com/huggingface/peft/blob/434651346ce138f67ebfe811764055975804c76f/src/peft/tuners/lora/dora.py#L89-L101

The only significant computation that could be cached is the `lora_weight` calculation. I did that and re-ran the benchmark, which resulted in:

```
eval: 10 runs, avg 0.1224 sec, min 0.1207 sec, max 0.1246 sec
train: 10 runs, avg 0.1553 sec, min 0.1541 sec, max 0.1563 sec
```

This is a decent speedup of ~20% but it still far from what we would get with just using LoRA.

**Comment by phemw** [comment]
Interesting, when you have the final weights, computing DoRA should be identical to LoRA except with an extra vector scaling operation, so I'm surprised the difference is only 20% (eval runtime isn't very close to LoRA). In the above code, even if you just returned ` base_result + lora_result * scaling`, I assume the difference would be relatively the same.

Is there a significant implementation difference between the two prior to this step?

**Comment by BenjaminBossan** [comment]
I noticed a flaw in my benchmark script. In that script, I alternate between benchmarking with `eval` and with `train` mode. But each time I switch to `train` mode, I clear the cache, making the next `eval` mode run slow. I fixed that and now run all `eval` mode runs first, then all `train` mode runs. The new results are:

```
eval:  10 runs, avg 0.0824 sec, min 0.0823 sec, max 0.0828 sec, total: 0.8243 sec
train: 10 runs, avg 0.1702 sec, min 0.1700 sec, max 0.1704 sec, total: 1.7017 sec
```

For comparison, here is LoRA:

```
eval:  10 runs, avg 0.0707 sec, min 0.0703 sec, max 0.0712 sec, total: 0.7067 sec
train: 10 runs, avg 0.0706 sec, min 0.0704 sec, max 0.0710 sec, total: 0.7065 sec
```

Overall, this is now only a 16% overhead for DoRA. This is all based on my PR #2661. So maybe you could give that branch a try on your own use case and report if you can confirm this speedup. Note that the PR is in an unfinished state and should only be used for testing. Also, caching will incur a memory overhead.

**Comment by BenjaminBossan** [comment]
_not stale_

**Comment by BenjaminBossan** [comment]
_not stale_

**Comment by BenjaminBossan** [comment]
_not stale, PR still in review_

**Comment by BenjaminBossan** [comment]
_not stale_

**Comment by BenjaminBossan** [comment]
_not stale_
